libraries/apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/ApolloCompilerPlugin.kt

```diff
@@ -65,8 +69,37 @@ interface ApolloCompilerPlugin {
   fun irOperationsTransform(): Transform<IrOperations>? {
     return null
   }
+
+  /**
+   * @return A list of [ForeignSchema] supported by this plugin
+   */
+  @ApolloExperimental
+  fun foreignSchemas(): List<ForeignSchema> {
+    return emptyList()
+  }
+
+  /**
+   * @return A [SchemaDocumentListener] called whenever the schema changed
+   */
+  @ApolloExperimental
+  fun schemaDocumentListener(): SchemaDocumentListener? {
+    return null
+  }
 }
 
+@ApolloExperimental
+interface SchemaDocumentListener {
+  /**
+   * Called when the schema changed and codegen needs to be updated
+   *
+   * @param schema the validated schema document.
+   * @param outputDirectory the compiler output directory. This directory is shared with the compiler, make sure to use a specific
+   * package name to avoid clobbering other files.
+   */
+  fun onSchemaDocument(schema: GQLDocument, outputDirectory: File)
```

Comment:
I left it out on purpose. The package name can be turing complete (because of `packageNameFromFilePath`, etc...). How big of an issue is it to not have it? If really needed, one can always pass it as a ACP (Apollo Compiler Plugin) argument.

Reply:
Ah I see. Not a big issue, but I think this means the package name will have to be repeated in most cases.

Reply:
Another question - why not passing the `Schema` rather than the `GQLDocument`? To correctly look at directives I should use `Schema.originalDirectiveName()` right?

Reply:
What about:

```kotlin
apollo {
  service("service") {
    packageName.set("com.example")
    plugin(project(":cache-plugin")) {
      argument("packageName", packageName.get())
    }
  }
}
```

And if evolution is a concern, hide it behind a java function:

```kotlin
apollo {
  service("service") {
    packageName.set("com.example")
    configureCachePlugin(this, project(":cache-plugin"))
  }
}

```

Reply:
>  why not passing the Schema rather than the GQLDocument? 

Not sure. GQLDocument is overall more stable than `Schema` but yea in your case, looks like you need `schema`. I'll amend.

Reply:
https://github.com/apollographql/apollo-kotlin/pull/6166 for passing a `Schema` instead of a `GQLDocument`

Reply:
`packageName.get()` is totally acceptable, a dedicated fun as well 👍.